File: memories/2024/exercise/java/day10/src/main/java/delivery/Building2.java
package delivery;

public class Building2 {
    
    private static final int ELF_UPWARD_MOVEMENT = 3;
    private static final int ELF_DOWNWARD_MOVEMENT = -2;
    private static final int STANDARD_UPWARD_MOVEMENT = 1;
    private static final int STANDARD_DOWNWARD_MOVEMENT = -1;
    private static final int SPECIAL_UPWARD_MOVEMENT = 42;
    private static final int SPECIAL_DOWNWARD_MOVEMENT = -2;
    private static final String ELF_INDICATOR = "🧝";

    public static int whichFloor(String instructions) {
        int currentFloor = 0;
        
        for (int position = 0; position < instructions.length(); position++) {
            char direction = instructions.charAt(position);
            currentFloor += calculateFloorChange(instructions, direction);
        }
        
        return currentFloor;
    }

    private static int calculateFloorChange(String instructions, char direction) {
        if (instructions.contains(ELF_INDICATOR)) {
            return direction == ')' 
                ? ELF_UPWARD_MOVEMENT 
                : ELF_DOWNWARD_MOVEMENT;
        } else if (!instructions.contains(ELF_INDICATOR)) {
            return direction == '(' 
                ? STANDARD_UPWARD_MOVEMENT 
                : STANDARD_DOWNWARD_MOVEMENT;
        } else {
            return direction == '(' 
                ? SPECIAL_UPWARD_MOVEMENT 
                : SPECIAL_DOWNWARD_MOVEMENT;
        }
    }
}

File: memories/2024/exercise/kotlin/day10/src/main/kotlin/delivery/Building2.kt
package delivery

/**
 * Building floor calculator for delivery instructions.
 */
object Building2 {
    private const val ELF_UPWARD_MOVEMENT = 3
    private const val ELF_DOWNWARD_MOVEMENT = -2
    private const val STANDARD_UPWARD_MOVEMENT = 1
    private const val STANDARD_DOWNWARD_MOVEMENT = -1
    private const val SPECIAL_UPWARD_MOVEMENT = 42
    private const val SPECIAL_DOWNWARD_MOVEMENT = -2
    private const val ELF_INDICATOR = "🧝"

    fun whichFloor(instructions: String): Int {
        var currentFloor = 0
        
        for (position in instructions.indices) {
            val direction = instructions[position]
            currentFloor += calculateFloorChange(instructions, direction)
        }
        
        return currentFloor
    }
    
    private fun calculateFloorChange(instructions: String, direction: Char): Int {
        return if (instructions.contains(ELF_INDICATOR)) {
            if (direction == ')') ELF_UPWARD_MOVEMENT else ELF_DOWNWARD_MOVEMENT
        } else if (!instructions.contains(ELF_INDICATOR)) {
            if (direction == '(') STANDARD_UPWARD_MOVEMENT else STANDARD_DOWNWARD_MOVEMENT
        } else {
            if (direction == '(') SPECIAL_UPWARD_MOVEMENT else SPECIAL_DOWNWARD_MOVEMENT
        }
    }
}